server/src/main/java/org/elasticsearch/tasks/TaskManager.java

```diff
@@ -495,6 +552,56 @@ public boolean hasParent(TaskId parentTaskId) {
         public CancellableTask getTask() {
             return task;
         }
+
+        synchronized void registerChildNode(DiscoveryNode node) {
+            if (banChildren) {
+                throw new TaskCancelledException("The parent task was cancelled, shouldn't start any children tasks");
+            }
+            if (childTasksPerNode == null) {
+                childTasksPerNode = new ObjectIntHashMap<>();
+            }
+            childTasksPerNode.addTo(node, 1);
+        }
+
+        void unregisterChildNode(DiscoveryNode node) {
+            final Runnable runnable;
+            synchronized (this) {
+                if (childTasksPerNode.addTo(node, -1) == 0) {
+                    childTasksPerNode.remove(node);
+                }
+                if (childTasksPerNode.isEmpty()) {
+                    runnable = onEmptyChildNodes;
+                } else {
+                    runnable = null;
```

Comment:
perhaps `runnable = {}`, which avoids the null check later down

Reply:
`onChildTasksCompleted` is still null if we do not cancel the task.